### ISSUE TITLE
Tsify doesn't generate typescript definitions under certain conditions

### ISSUE THREAD

**Issue by bes** [madonoharu/tsify#50]
# Bug

Sometimes tsify won't generate a TypeScript interface.

# How to reproduce

Specifically, under these circumstances:

Given a crate that has _exactly one_ type that has a Tsify derive, and `into_wasm_abi` but NOT `from_wasm_abi`: 

`crates/somecrate/lib.rs`
```rust
#[derive(Clone, Debug)]
#[derive(tsify::Tsify, serde::Serialize)]
#[tsify(into_wasm_abi)]
pub struct ExampleType {
    pub name: Option<String>,
}
```

AND the type _does not_ appear in a direct return value (but it does appear in an indirect return value)

`my_wasm_lib/lib.rs`
```rust
#[wasm_bindgen(js_class = ExampleTypeHandleJs)]
impl ExampleTypeHandle {
    #[wasm_bindgen(js_name = getExampleTypeVec)]
    pub fn get_example_type_vec(&self) -> Option<ExampleTypeVec> {
        todo!()
    }
}

#[derive(Tsify, Serialize)]
#[tsify(into_wasm_abi)]
pub struct ExampleTypeVec(pub Vec<ExampleType>);
```

Then there will be no `interface ExampleType` in the resulting xxx.d.ts file.

# Mitigation

However, this can be mitigated by one or more of the following:

## Let the type appear as a _direct_ return type

```rust
    #[wasm_bindgen(js_name = getExampleType)]
    pub fn get_example_type(&self) -> ExampleType {
        todo!()
    }
```

This is solution is unfortunate, because I have no use for such a method.

## Let the type use `from_wasm_abi`

```rust
#[derive(Clone, Debug)]
#[derive(tsify::Tsify, serde::Serialize, serde::Deserialize)]
#[tsify(into_wasm_abi, from_wasm_abi)]
pub struct ExampleType {
    pub name: Option<String>,
}
```

This solution is unfortunate, because I have no use for `from_wasm_abi` and `serde::Deserialize` which just add bloat.

@siefkenj I got a tip from @cwfitzgerald that you have a new fork with recente fixes, would you be open to enabling issues on your fork to get stuffed fixed there?

**Comment by siefkenj** [comment]
@bes Yes, I would be open to a PR on https://github.com/siefkenj/tsify  Please make your PR to the `next` branch.

**Comment by bes** [comment]
Curiously, there is also a third mitigation:

# Mitigation

## Compile the wasm-library with `codegen-units = 1`

When compiling in release mode, the bug does not appear, which made me interested to find out why. I tried compiling dev mode with `codegen-units = 1` which also made the TypeScript interface show up.

**Comment by CinchBlue** [comment]
I believe we just encountered this -- seems to still be present.

**Comment by bes** [comment]
For me this is caused because of a multi-crate setup where some types do not survive linking. I created an issue in the wasm-bindgen repository to start a discussion to see if there is something that can be done. See the linked issue above.

**Comment by madonoharu** [comment]
Your follow-up in [rustwasm/wasm-bindgen#5137](https://github.com/rustwasm/wasm-bindgen/issues/5137) nailed the diagnosis: the declaration isn't skipped at generation, it's dropped at link time, and `codegen-units = 1` brings it back.

tsify emits the declaration inside a `#[wasm_bindgen(typescript_custom_section)]` const. As I understand it, wasm-bindgen turns that into a custom section in the compiled `.wasm` that its CLI collects afterward, and a type referenced only indirectly can land in a codegen unit whose section doesn't survive that step, taking the declaration with it. The drop happens at link time, long after the macro has run — tsify never sees it happen.

Workaround for now:

```toml
[profile.release]
codegen-units = 1
```

I'm keeping this open instead of closing it as strictly wasm-bindgen's problem. #5137 still has no maintainer response, and whether tsify can do anything on its end — forcing the linker to keep that section, for instance — isn't ruled out yet.

Sorry this took so long to get back to, and thanks for chasing it upstream and linking it back here — that saved a lot of guesswork.
